index/scorch/segment/int.go

```diff
@@ -92,3 +92,67 @@ func DecodeUvarintAscending(b []byte) ([]byte, uint64, error) {
 	}
 	return b[length:], v, nil
 }
+
+// ------------------------------------------------------------
+
+type MemUvarintReader struct {
+	C int64 // index of next byte to read from S
+	S []byte
+}
+
+func NewMemUvarintReader(s []byte) *MemUvarintReader {
+	return &MemUvarintReader{S: s}
+}
+
+// Len returns the number of unread bytes.
+func (r *MemUvarintReader) Len() int {
+	n := int(int64(len(r.S)) - r.C)
+	if n < 0 {
+		return 0
+	}
+	return n
+}
+
+// ReadUvarint reads an encoded uint64.
+func (r *MemUvarintReader) ReadUvarint() uint64 {
```

Comment:
Could I persuade you to make the return type (uint64, error) and return an error instead of panicking?

It would make it more consistent with other readers, and it would allow for the possibility of handling a corrupted index in some way that avoids the whole process exiting (like what if your other indexes are fine, and hot in memory right now?)

Reply:
totally reasonable -- will give it a shot and see if it affects perf.